source/languageDetection.py

```diff
@@ -0,0 +1,284 @@
+
+def updateLanguagePriorityFromConfig():
+	"""read string from config and convert it to list"""
+	global languagePriorityListSpec 
+	tempList = []
+	try:
+		languageList = config.conf["languageDetection"]["preferredLanguages"]
+		for language in languageList: 
+			tempList.append( [ language , getScriptIDFromLangID(language) , getLanguageDescription( language ) ]) 
+		languagePriorityListSpec = tempList 
+	except KeyError:
+		pass
+
+langIDToScriptID = OrderedDict([
+	("en" , "Latin"),
+	("af_ZA" , "Latin"),
+	("ca" , "Latin"),
+	("cs" , "Latin"),
+	("da" , "Latin"),
+	("de" , "Latin"),
+	("el" , ("Latin" , "Greek" ) ),
+	("es" , "Latin"),
+	("fi" , "Latin"),
+	("fr" , "Latin"),
+	("ga" , "Latin"),
+	("gl" , "Latin"),
+	("hr" , "Latin"),
+	("nl" , "Latin"),
+	("pl" , "Latin"),
+	("pt_br" , "Latin"),
+	("sl" , "Latin"),
+	("am" , "Armenian"),
+	("ar" , "Arabic"),
+	("as" , "Bengali"),
+	("bg" , "Cyrillic"),
+	("bn" , "Bengali"),
+	("fa" , "Arabic"),
+	("gu" , "Gujarati"),
+	("ka" , "Georgian"),
+	("kn" , "Kannada"),
+	("mk" , "Cyrillic"),
+	("ml" , "Malayalam"),
+	("mn" , "Mongolian"),
+	("he" , "Hebrew"),
+	("hi" , "Devanagari"),
+	("mr" , "Devanagari"),
+	("ne" , "Devanagari"),
+	("sa" , "Devanagari"),
+	("or" , "Oriya"),
+	("pa" , "Gurmukhi"),
+	("ru" , "Cyrillic"),
+	("sq" , "Caucasian_Albanian"),
+	("ta" , "Tamil"),
+	("te" , "Telugu"),
+	("sr" , ("Latin" , "Cyrillic")),
+	("ja" , ("Han", "Hiragana", "Katakana", "FullWidthNumber" )), 
+	("ko" , ("Han", "Hiragana", "Katakana", "Hangul" , "FullWidthNumber" )), 
+	("zh" , ("Han", "Hiragana", "Katakana", "FullWidthNumber"  )), 
+])
+
+def getLanguagesWithDescriptions(ignoreLanguages=None):
+	"""generates a list of locale names, plus their full localized language and country names.
+	@rtype: list of tuples
+	"""
+	#Make a list of all the languages found for language to script mapping
+	allLanguages  = langIDToScriptID.keys()
+	allLanguages.sort()
+	languageDescriptions = []
+	if ignoreLanguages is None:
+		ignoreLanguages = {lang[0] for lang in languagePriorityListSpec}
+	for language in allLanguages:  
+		if language in ignoreLanguages: 
+			continue
+		else:
+			desc=languageHandler.getLanguageDescription(language )
+			label="%s, %s"%(desc,language ) if desc else language 
+			languageDescriptions.append((language , label))
+	return languageDescriptions
+
+def getScriptCode(chr):
+	"""performs a binary search in scripCodes for unicode ranges
+	@param chr: character for which a script should be found
+	@type chr: string
+	@return: script code
+	@rtype: int"""
+	mStart = 0
+	mEnd = len(scriptCode)-1
+	characterUnicodeCode = ord(chr)
+	# Number should respect preferred language setting
+	# FullWidthNumber is in Common category, however, it indicates Japanese language context
+	if 0x30 <= characterUnicodeCode <= 0x39:
+		return "Number"
+	elif 0xff10 <= characterUnicodeCode <= 0xff19:
+		return "FullWidthNumber"
+	while( mEnd >= mStart ):
+		midPoint = (mStart + mEnd ) >> 1
```

Comment:
Is there not a built in way of performing this search? So we don't reinvent the wheel, so to speak?

Reply:
I don't think so as is it not a simple binary search.

Reply:
You could use `bisect`, in the initialisation create a list of the `rangeEndValues` like so:
`unicodeScriptRangeEnd = [ k[1] for k in scriptCode]`
Then in this function you can do:
```
# Based on the following assumptions: 
# - ranges must overlap
# - range end and start values are included in that range
# - there may be gaps between ranges.

# Approach: Look for the first index of a range where the range end value is greater
# than the code we are searching for. If this is found, and the start value for this range
# is less than or equal to the code we are searching for then we have found the range.
# That is startValue <= characterUnicodeCode <= endValue

index = bisect.bisect_left(unicodeScriptRangeEnd, characterUnicodeCode )
if index == len(unicodeScriptRangeEnd):
    # there is no value of index such that: `characterUnicodeCode <= scriptCode[index][1]`
    # characterUnicodeCode is larger than all of the range end values so a range is not 
    # found for the value:
    return None

# Since the range at index is the first where `characterUnicodeCode <= rangeEnd` is True,
# we now ensure that for the range at the index `characterUnicodeCode >= rangeStart` 
# is also True. 
candidateRange = scriptCode[index]
rangeStart = candidateRange[0]
if rangeStart > characterUnicodeCode :
    # characterUnicodeCode comes before the start of the range at index so a range 
    # is not found for the value
    return None
rangeName = candidateRange[2]
return rangeName
```

Reply:
The assumption of ranges overlapping is not valid. 

At the outset this algorithm looks to me is O(N) and binary search is O(log N) n being the number of ranges.
The loop 
unicodeScriptRangeEnd = [ k[1] for k in scriptCode]
runs for number of ranges.
getScriptCode is called for every character so it should be as fast as possible. Considering that I think binary search is better.

I am not familiar with bisect so I am looking in to it.
Isn't it better to not modify this function if it is working fine?

Reply:
> The assumption of ranges overlapping is not valid.

Oh, the comment meant to say "must not overlap"

> At the outset this algorithm looks to me is O(N) and binary search is O(log N) n being the number of ranges.

I believe [bisect](https://docs.python.org/2/library/bisect.html) is O(log N). There is the step to split out the rangeEndValues, though this should only be done once, and as such should not be noticable. If the performance of the function is such a concern, then we should measure it and compare results of optimised results.

> Isn't it better to not modify this function if it is working fine?
My concerns were readability, and for edge cases that are easy to miss, and therefor write tests for.

Reply:
As I said earlier, bisect would be O(log N) but the loop before that would be O(N). We can take that loop out, but that method looks more complicated to me.
unicodeScriptRangeEnd = [ k[1] for k in scriptCode]

I would try to write test than change this algorithm which took me some time to test. But at that time, we didn't have unit tests in NVDA so I had only done manual testing.

Reply:
Yes, the creation of `unicodeScriptRangeEnd` is O(N). But it is only done once, so it does not need to be considered. Since speed is a concern I thought I would measure these approaches, see [this gist with a timing script](https://gist.github.com/feerrenrut/f66ac4b145ee0d435e557d3361b92440). Place the file `testSpeedUnicodeCharacterLookup.py` in the `<nvdaRepoRoot>/source/` directory of your branch.

The results on my machine were:
```
time python testSpeedUnicodeCharacterLookup.py
using 100 iterations
testing over range 0-65535, a total of 65535 values
withBisect: 5.69379344301
customBinarySearch: 12.9052367034

real    0m18.679s
user    0m0.015s
sys     0m0.000s
```

I have not profiled the two solutions, so I can not say why bisect is so much faster.

Reply:
You will also notice in this script that there are problems with using `ord()` on values of `0x10000` / `65536` or greater. This is something that might need to be considered for this PR. Is there any situation where `chr` passed to `getScriptCode()` could be this large? The values in `scriptRanges` go up to `0Xe007f `.

Reply:
Thanks for your time check. I have implemented your suggestion and removed my custom implementation of binary search.